README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SurveySolutionsAPI

<!-- badges: start -->

[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-green.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![R-CMD-check](https://github.com/michael-cw/SurveySolutionsAPI/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/michael-cw/SurveySolutionsAPI/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

This is the development version of the [Survey Solutions R API
package](https://docs.mysurvey.solutions/headquarters/api/api-r-package/).
It allows you to export data/paradata collected through CAWI, CATI or
CAPI operations, manipulate users and questionnaires, create
assignments, get information about the survey progress etc.. All
directly out of R.

However it is more than just a simple wrapper around the Survey
Solutions RESTful API, which one can easily address through using
i.e. the excellent httr package which also constitutes the basis for
this package. The main intention of this package though, is to be
integrated into your own data collection workflow, either through a
shiny application or just through a simple script. This allows you to
build your own workflows with your own customized user interfaces for
your Survey Solutions data collection operations, no matter if you are
dealing with a small scale impact evaluation or a large scale census.
For this reason, the API calls return easy to process data tables, and
in some cases, like paradata, even add additional variables useful for
further processing.

Important to note here is, that the package also makes extensive use of
the **data.table** package, which allows for fast (multi-core)
processing of large data sets. The individual functions either return a
data.table, or a specific S3 extension of data.table, with individual
methods defined. In case you are not familiar with it yet, you should
probably have a look a this
[introduction](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-intro.html).

Nevertheless, in case you don’t require the capacities offered by
data.table, you can still treat the return objects like a data frame or
even continue with it in the tidyverse. A package which combines both
worlds is the [dtplyr](https://github.com/tidyverse/dtplyr) package, and
for a comparison between the dplyr package and the data.table package
see [here](https://atrebas.github.io/post/2019-03-03-datatable-dplyr/).

Specific changes in version 0.6.X:

  - Preparation for token authentication (not active yet)
  - Map handling and management: suso\_mapupload(), suso\_mapassign(),
    suso\_mapinfo()
  - Introduction of S3 classes for return values, such that all
    information returned from the API can be used efficiently. See news
    for further details and classes.

## Installation

The package is not on CRAN yet, so you have to install it by using

``` r

devtools::install_github("michael-cw/SurveySolutionsAPI", build_vignettes = T)
```

The included vignettes also contain different examples and further
explanations. Bellow you can find a few examples.

## Examples

### Credentials Management

To use the API you first need to set-up the API user on your Survey
Solutions server. See
[here](https://support.mysurvey.solutions/headquarters/api/) for
details. After this done, you can use the *suso\_set\_key()* function,
to provide your credentials.

``` r
library(SurveySolutionsAPI)
suso_clear_keys()
suso_set_key("https://xxx.mysurvey.solutions", "xxxxxx", "xxxxxxx")
suso_keys()
#> $suso
#> $suso$susoServer
#> [1] "https://xxx.mysurvey.solutions"
#> 
#> $suso$susoUser
#> [1] "xxxxxx"
#> 
#> $suso$susoPass
#> [1] "xxxxxxx"
#> 
#> 
#> attr(,"class")
#> [1] "suso_api"
```

### Workspace Management

You can also manage your workspaces through the API. In particular, you
can

  - request a list of all workspaces the API user has access to, as well
    as
  - create new workspaces
  - and assign users to a particular workspace.

However for the latter two operations you require ADMIN credentials.
Standard API user credentials would not work.

To receive a list of workspaces to which the current API user has access
to, you can just run:

``` r
ws <- suso_getWorkspace()
print(head(ws))
#>       Name       DisplayName DisabledAtUtc
#> 1: primary Default Workspace            NA
#> 2:    test              test            NA
#> 3:   test2             test2            NA
```

### User Management

These functions are particularly useful for survey management, and more
details can be found in the corresponding vignette on survey management.
Lets start with getting a list of all supervisors on the server.

``` r
sv <- suso_getSV(workspace = "test")
print(head(sv))
#>    IsLocked        CreationDate                               UserId   UserName
#> 1:    FALSE 2023-02-07 17:33:40 00d21677-0331-4698-ae7b-908aeac14dd6 somesvuser
#> 2:    FALSE 2023-02-04 19:05:10 ded27a4a-716d-4bf2-8d54-aedc31fc164b     SV0001
#> 3:    FALSE 2023-02-04 19:36:03 5753ddde-3c7d-438f-9934-43fdf18fa802     SV0002
#> 4:    FALSE 2023-02-04 19:36:03 72196c84-8f2f-488a-8d61-f23b8b998d43     SV0003
#> 5:    FALSE 2023-02-04 19:36:03 1a9d4b10-c92d-401d-af69-2a7d0a391fe6     SV0004
#> 6:    FALSE 2023-02-04 19:36:03 97e9e4ff-22e1-492f-a102-03c2115903e7     SV0005
```

### Questionnaire Management

The basic questionnaire API calls are handled through the
*suso\_getQuestDetails* function.

If no input is provided, the function returns a list of all
questionnaires on the server:

``` r
questlist <- suso_getQuestDetails(workspace = "test")
# print(questlist)
```

Specifying *operation.type = status*, you receive a list of
statuses.

``` r
statlist <- suso_getQuestDetails(operation.type = "statuses", workspace = "test")
print(statlist)
#>  [1] "Restored"               "Created"                "SupervisorAssigned"     "InterviewerAssigned"    "RejectedBySupervisor"   "ReadyForInterview"      "SentToCapi"            
#>  [8] "Restarted"              "Completed"              "ApprovedBySupervisor"   "RejectedByHeadquarters" "ApprovedByHeadquarters" "Deleted"
```

By taking a particular *QuestionnaireId* and specifying the
*operation.type *you can execute further requests. For example,

### Data Export

To export the data collected in Survey Solutions, you use
*suso\_export*.

    #> The last file has been created 0.06833 hours ago.
    #> FileName: assignment__actions 
    #> Nesting Level: 0 
    #> ****
    #> 
    #> FileName: households 
    #> Nesting Level: 1 
    #> ****
    #> 
    #> FileName: interview__actions 
    #> Nesting Level: 0 
    #> ****
    #> 
    #> FileName: interview__diagnostics 
    #> Nesting Level: 0 
    #> ****
    #> 
    #> FileName: interview__errors 
    #> Nesting Level: 0 
    #> ****

Its return value is a list with the following elememts: main, R1, R2,
R3, with

  - **main** containing the files: interview\_\_comments,
    kasai\_listing1
  - **R1** containing all rosters at the first level
  - **R2** containing all rosters at the second level
  - **R3** containing all rosters at the third level

through the harmonized ID, main and roster files can easily be put
together. More on this in the specific vignette.

## Paradata Export

To retrieve the paradata for a particular interview you use
*suso\_export\_paradata*

``` r
system.time(para1 <- suso_export_paradata(questID = questlist[1, QuestionnaireId], workspace = "test", version = questlist[1, Version], reloadTimeDiff = 24, onlyActiveEvents = F, allResponses = T))
#> 
#> The last file has been created 1 hours ago.
#> 
#> Starting download & file extraction. 
#> 
#> 
#> Calculating Response Timings.
#> 
#> Extracting GPS variable.
#> Processing: 
#>  AnswerSet
#> 
#>  AnswerRemoved
#> 
#>  ApproveByHeadquarter
#> 
#>  Restarted
#> 
#>  Reject
#> 
#>  QuestionDeclaredInvalid
#> 
#>  QuestionDeclaredValid
#> 
#> Export & Transformation finished.
#>    user  system elapsed 
#>   0.536   0.036   0.820
```

### Map management

Since recently, it is now also possible, to handle maps through the API.
This is implemented through the lately introduced GraphQL API.

To upload a map, you can just use the **suso\_mapupload** function like
this:

``` r
suso_mapupload(workspace = "test", path_to_zip = mapPath)
#>    xMaxVal yMaxVal xMinVal yMinVal wkid             fileName  size maxScale minScale shapeType       importDateUtc                           uploadedBy     users
#> 1:  -78.39    26.8  -78.62   26.55 4326 checkshapesimple.shp 20467        0        0   Polygon 2023-02-16 17:24:47 961d073f-f316-4353-9e7e-1f00899cb837 <list[0]>
```

To assign a map you can use the **suso\_mapassign**
function:

``` r
suso_mapassign(workspace = "test", fileName = "checkshapesimple.shp", userName = "INT0004")
#>                fileName    user shapeType       importDateUtc
#> 1: checkshapesimple.shp INT0004   Polygon 2023-02-16 17:24:47
```

## Further information

To find information on the World Bank’s Survey Solutions CASS have a
look on these pages:

  - Survey Solutions Support Page: <https://support.mysurvey.solutions/>

  - Survey Solutions Server request: <https://mysurvey.solutions/>

To find information on the API syntax, visit your own servers API
documentation or got to:

  - <https://demo.mysurvey.solutions/primary/apidocs/index.html#>

And to find more information on the package, check out the package
vignettes.

## Feature requests and bug reports

You can either use the standard GitHub approach by filing a bug
report/feature request
[here](https://github.com/michael-cw/SurveySolutionsAPI/issues) or you
use the Survey Solutions user forum
[here](https://forum.mysurvey.solutions/c/api/13).


NEWS.md
# SurveySolutionsAPI 0.6.4

* Added S3 *assignmentClass*.
* Added dedicated vignette for assignments


# SurveySolutionsAPI 0.6.3

* Added *suso_patchApproveInterview*, *suso_mapinfo*, *suso_mapassign* and *suso_gql_uploadmap*.
* Updated *suso_patchRejectInterview* to work with re-assignment of interview.



# SurveySolutionsAPI 0.6.2

* Added start and end date to *suso_getINT_log*.


# SurveySolutionsAPI 0.6.1

* Updated *suso_export* and *suso_export_paradata* to use the same in *inShinyApp* parameter.


# SurveySolutionsAPI 0.6.0

* Updated all workspace functions.




R/suso_api_Interviews.R
#' Survey Solutions API call to retrieve all interviews for a specific questionnaire
#'
#' Returns all interviews for the specified questionnaire and the selected status.
#'
#' @details ATTENTION: This function will soon be switched to the GraphQL endpoint, since the REST API is
#' already deprecated.
#'
#'
#' @param server Survey Solutions server address
#' @param apiUser Survey Solutions API user
#' @param apiPass Survey Solutions API password
#' @param workspace server workspace, if nothing provided, defaults to primary
#' @param token If Survey Solutions server token is provided \emph{apiUser} and \emph{apiPass} will be ignored
#' @param questID your Survey Solutions \emph{QuestionnaireId}. Retrieve a list of questionnaires by executing \code{suso_getQuestDetails}
#' @param version version of the questionnaire
#' @param workStatus define which statuses the file should inlude (i.e. \emph{Restored,Created,SupervisorAssigned,InterviewerAssigned,
#' RejectedBySupervisor,ReadyForInterview,
#' SentToCapi,Restarted,Completed,ApprovedBySupervisor,
#' RejectedByHeadquarters,ApprovedByHeadquarters,Deleted}), if NULL only completed interviews will be shown.
#'
#'
#' @examples
#' \dontrun{
#' To get all interviews for a specific questionnaire and a specific status
#'      suso_getAllInterviewQuestionnaire(
#'               workspace = "myworkspace",
#'               questID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea",
#'               version = 1,
#'               workStatus = "InterviewerAssigned"
#'               )
#' # or to get all interviews in status completed
#'      suso_getAllInterviewQuestionnaire(
#'               workspace = "myworkspace",
#'               questID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea",
#'               version = 1
#'               )
#'
#' }
#'
#'
#' @export
#' @importFrom lubridate as_datetime

suso_getAllInterviewQuestionnaire <- function(server= suso_get_api_key("susoServer"),
                                              apiUser=suso_get_api_key("susoUser"),
                                              apiPass=suso_get_api_key("susoPass"),
                                              workspace = NULL,
                                              token = NULL,
                                              questID = "",
                                              version = 1,
                                              workStatus = "Completed") {
    ######################################
    # Check arguments
    #  - workStatus
    margs<-suso_getQuestDetails(operation.type = "statuses", workspace = workspace)
    if(!is.null(workStatus)) {
        workStatus<-match.arg(workStatus, margs)
    } else {
        workStatus<-"Completed"
    }

    ## workspace default
    workspace<-.ws_default(ws = workspace)
    ## Define API
    server<-httr::parse_url((server))
    server$scheme<-"https"
    server$path<-file.path(workspace,"api", "v1", "interviews", fsep = "/")
    server$query<-list(questionnaireId = questID,
                       questionnaireVersion = version,
                       status = workStatus,
                       pageSize = 40,
                       page = 1)

    ## authentication
    auth<-authenticate(apiUser, apiPass, type = "basic")

    ## request
    test_detail <- GET(url = build_url(server),
                       auth,
                       httr::content_type_json(),
                       httr::user_agent("r api v1"),
                       httr::accept_json())
    check_response(test_detail)

    aJsonFile <- tempfile()
    writeBin(content(test_detail, "raw"), aJsonFile)
    test_json <- fromJSON(aJsonFile)
    # stop if empty response
    if (length(test_json$Interviews)==0) stop("No data for this questionnaire. Did you use the correct ID?", call. = F)
    qTot <- test_json$TotalCount
    # 2. Check if more than 40, append rest
    if (qTot > nrow(test_json$Interviews)) {
        qTotRest <- qTot
        repCalls <- ceiling(qTotRest/nrow(test_json$Interviews))
        for (i in 2:repCalls) {
            test_detail <- GET(url = modify_url(server, query = c(server$query, page = i)),
                               auth,
                               httr::content_type_json(),
                               httr::user_agent("r api v1"),
                               httr::accept_json())
            check_response(test_detail)
            aJsonFile <- tempfile()
            writeBin(content(test_detail, "raw"), aJsonFile)
            test_json_tmp <- fromJSON(aJsonFile)
            test_json$Interviews <- rbind(test_json$Interviews, test_json_tmp$Interviews)
        }
    }
    # Only data.table of interviews is returned
    test_json<-data.table(test_json$Interviews)
    # Set date time to utc with lubridate
    if(nrow(test_json)>0) test_json[,LastEntryDate:=lubridate::as_datetime(LastEntryDate)][]
    return(test_json)
}

#' Survey Solutions API call to retrieve all answers for a specific interview
#'
#' Returns all responses for a specific interview
#'
#' @param server Survey Solutions server address
#' @param apiUser Survey Solutions API user
#' @param apiPass Survey Solutions API password
#' @param workspace server workspace, if nothing provided, defaults to primary
#' @param token If Survey Solutions server token is provided \emph{apiUser} and \emph{apiPass} will be ignored
#' @param intID the \emph{InterviewId} of the interview. To get a list of all interview for a specific questionnaire, execute \code{suso_getAllInterviewQuestionnaire}
#'
#' @examples
#' \dontrun{
#' suso_getAllAnswerInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea"
#'           )
#'
#' }
#'
#'
#' @export
#'
suso_getAllAnswerInterview <- function(server= suso_get_api_key("susoServer"),
                                       apiUser=suso_get_api_key("susoUser"),
                                       apiPass=suso_get_api_key("susoPass"),
                                       workspace = NULL,
                                       token = NULL,
                                       intID = "") {
    ## workspace default
    workspace<-.ws_default(ws = workspace)
    ## Define API
    server<-httr::parse_url(paste0(server))
    server$scheme<-"https"
    server$path<-file.path(workspace,"api", "v1", "interviews", intID, fsep = "/")
    server<-build_url(server)
    ## authentication
    auth<-authenticate(apiUser, apiPass, type = "basic")
    ## Call
    test_detail <- GET(url = server,
                       auth,
                       httr::content_type_json(),
                       httr::user_agent("r api v1"),
                       httr::accept_json())
    check_response(test_detail)
    aJsonFile <- tempfile()
    writeBin(content(test_detail, "raw"), aJsonFile)
    test_json <- fromJSON(aJsonFile)
    # Export only answers: Not data.table!
    test_json<-data.table(test_json$Answers)
    return(test_json)
}

#' Get all history for a specific interview
#'
#' @param server Survey Solutions server address
#' @param apiUser Survey Solutions API user
#' @param apiPass Survey Solutions API password
#' @param workspace server workspace, if nothing provided, defaults to primary
#' @param token If Survey Solutions server token is provided \emph{apiUser} and \emph{apiPass} will be ignored
#' @param intID the \emph{InterviewId} of the interview.
#'
#' @examples
#' \dontrun{
#' suso_getAllHistoryInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea"
#'           )
#'
#' }
#'
#' @export
suso_getAllHistoryInterview <- function(server= suso_get_api_key("susoServer"),
                                        apiUser=suso_get_api_key("susoUser"),
                                        apiPass=suso_get_api_key("susoPass"),
                                        workspace = NULL,
                                        token = NULL,
                                        intID = "") {
    ## workspace default
    workspace<-.ws_default(ws = workspace)
    ## Define API
    server<-httr::parse_url(paste0(server))
    server$scheme<-"https"
    server$path<-file.path(workspace,"api", "v1", "interviews", intID, "history")
    ## authentication
    auth<-authenticate(apiUser, apiPass, type = "basic")

    ## Call
    test_detail <- GET(url = build_url(server),
                       auth,
                       httr::content_type_json(),
                       httr::user_agent("r api v1"),
                       httr::accept_json())
    check_response(test_detail)
    aJsonFile <- tempfile()
    writeBin(content(test_detail, "raw"), aJsonFile)
    test_json <- fromJSON(aJsonFile)
    # Export only records
    test_json<-data.table(test_json$Records)
    # Set date time to utc with lubridate
    test_json[,Timestamp:=as_datetime(Timestamp)][]
    return(test_json)
}


#' Get statistics for interview
#'
#' @description Fetch statistics for a specific interview, vectorized over interview id (int_id)
#'
#' @param server Survey Solutions server address
#' @param apiUser Survey Solutions API user
#' @param apiPass Survey Solutions API password
#' @param workspace server workspace, if nothing provided, defaults to primary
#' @param token If Survey Solutions server token is provided \emph{apiUser} and \emph{apiPass} will be ignored
#' @param intID a single or multiple \emph{InterviewId}.
#'
#' @examples
#' \dontrun{
#' suso_get_stats_interview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea"
#'           )
#'
#' }
#'
#' @export



suso_get_stats_interview<-function(server= suso_get_api_key("susoServer"),
                                   apiUser=suso_get_api_key("susoUser"),
                                   apiPass=suso_get_api_key("susoPass"),
                                   workspace = NULL,
                                   token = NULL,
                                   intID = "") {
    ## workspace default
    workspace<-.ws_default(ws = workspace)
    ## Define API
    server<-httr::parse_url(paste0(server))
    server$scheme<-"https"
    ## authentication
    auth<-authenticate(apiUser, apiPass, type = "basic")

    tj<-list()
    for(id in intID){
        server$path<-NULL
        server$path<-file.path(workspace,"api", "v1", "interviews", id, "stats", fsep = "/")
        ## Call
        test_detail <- GET(url = build_url(server),
                           auth,
                           httr::content_type_json(),
                           httr::user_agent("r api v1"),
                           httr::accept_json())
        if (status_code(test_detail)!=200){
            warning(paste0("No data for interview: " , id,". Did you specify the correct ID?"), call. = F)
            print(id)
            next()
        }
        aJsonFile <- tempfile()
        writeBin(content(test_detail, "raw"), aJsonFile)
        test_json <- fromJSON(aJsonFile)
        test_json<-as.data.table(t(unlist(test_json)))
        tj[[id]]<-test_json
    }
    tj<-rbindlist(tj, fill = T)
    # to numeric conversion of counts
    names_col<-names(tj)[c(1:9,15:17)]
    for (col in names_col) set(tj, j=col, value=as.numeric(tj[[col]]))
    ## date conversion
    tj[,UpdatedAtUtc:=as_datetime(UpdatedAtUtc)]
    tj[,InterviewDuration:=as.POSIXct(InterviewDuration, format = "%H:%M:%OS")][]
    return(tj)
}

#' Reject interviews either as supervisor or as headquarter.
#'
#' @description Allows you to reject interviews in supervisor or headquarters role
#' as well as to provide a comment (i.e. reason) for the rejection.
#'
#'
#' @details  For details please see:
#' \url{https://docs.mysurvey.solutions/headquarters/interviews/survey-workflow/}
#'
#'
#' @param server Survey Solutions server address
#' @param apiUser Survey Solutions API user
#' @param apiPass Survey Solutions API password
#' @param workspace server workspace, if nothing provided, defaults to primary
#' @param token If Survey Solutions server token is provided \emph{apiUser} and \emph{apiPass} will be ignored
#' @param intID the \emph{InterviewId} of the interview.
#' @param new_uid if provided the interview will be rejected to a different user than the current one.
#' @param HQ if FALSE, reject as supervisor, if TRUE rejected as headquarters
#' @param comment comment which should be sent with the questionnaire
#'
#' @examples
#' \dontrun{
#' # reject the interview as supervisor
#' suso_patchRejectInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea"
#'           )
#' # reject the interview as headquarters
#' suso_patchRejectInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea",
#'           HQ = TRUE
#'           )
#' # reject the interview and provide a comment, so the interviewer knows about the problem
#' suso_patchRejectInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea",
#'           comment = "Too many errors, please check and re-submit!"
#'           )
#' # reject the interview to a different user
#' suso_patchRejectInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea",
#'           new_uid = "3b0c6e09-d606-4914-9e20-2abc048d5bea"
#'           )
#'
#' }
#'
#' @export
#'
suso_patchRejectInterview <- function(server= suso_get_api_key("susoServer"),
                                      apiUser=suso_get_api_key("susoUser"),
                                      apiPass=suso_get_api_key("susoPass"),
                                      workspace = NULL,
                                      token = NULL,
                                      intID = "",
                                      new_uid = NULL,
                                      HQ = FALSE,
                                      comment = "Please check errors and re-submit!") {
    ## select reject
    reject<-ifelse(HQ, "hqreject", "reject")
    ## workspace default
    workspace<-.ws_default(ws = workspace)
    ## Define the api
    url <- parse_url(server)
    url$scheme <- "https"
    url$path<-file.path(workspace,"api", "v1", "interviews", intID, reject, fsep = "/")
    url$query <- list(comment = comment, responsibleId = new_uid)
    ## authentication
    auth<-authenticate(apiUser, apiPass, type = "basic")

    server <- build_url(url)
    test_detail <- PATCH(url = server, auth)
    if(test_detail$status_code==406) {
        test_json<-data.table::data.table(resp_status=test_detail$status_code, intID=intID, new_uid=new_uid, description="incorrect status")
        return(test_json)
    } else if(test_detail$status_code==404) {
        test_json<-data.table::data.table(resp_status=test_detail$status_code, intID=intID, new_uid=new_uid, description="interview not found")
        return(test_json)

    } else if(test_detail$status_code==200) {
        test_json<-data.table::data.table(resp_status=test_detail$status_code, intID=intID, new_uid=new_uid, description="success")
        return(test_json)

    } else {
        test_json<-data.table::data.table(resp_status=test_detail$status_code, intID=intID, new_uid=new_uid, description="other error")
        return(test_json)
    }
}

#' Approve interviews either as supervisor or as headquarter.
#'
#' @description Allows you to approve interviews in supervisor or headquarters role
#' as well as to provide a comment (i.e. reason) for the approval.
#'
#'
#' @details  For details please see:
#' \url{https://docs.mysurvey.solutions/headquarters/interviews/survey-workflow/}
#'
#'
#' @param server Survey Solutions server address
#' @param apiUser Survey Solutions API user
#' @param apiPass Survey Solutions API password
#' @param workspace server workspace, if nothing provided, defaults to primary
#' @param token If Survey Solutions server token is provided \emph{apiUser} and \emph{apiPass} will be ignored
#' @param intID the \emph{InterviewId} of the interview.
#' @param HQ if FALSE, approve as supervisor, if TRUE rejected as headquarters
#' @param comment comment which should be sent with the questionnaire
#'
#' @examples
#' \dontrun{
#' # approve the interview as supervisor
#' suso_patchApproveInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea"
#'           )
#' # approve the interview as headquarters
#' suso_patchApproveInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea",
#'           HQ = TRUE
#'           )
#' # approve the interview and provide a comment
#' suso_patchApproveInterview(
#'           workspace = "myworkspace",
#'           intID = "dee7705f-d611-4b12-9b97-2b8e5b80c4ea",
#'           comment = "Well done!"
#'           )
#'
#' }
#'
#' @export
#'
suso_patchApproveInterview <- function(server= suso_get_api_key("susoServer"),
                                      apiUser=suso_get_api_key("susoUser"),
                                      apiPass=suso_get_api_key("susoPass"),
                                      workspace = NULL,
                                      token = NULL,
                                      intID = "",
                                      HQ = FALSE,
                                      comment = "Well done!") {
    ## select reject
    approve<-ifelse(HQ, "hqapprove", "approve")
    ## workspace default
    workspace<-.ws_default(ws = workspace)
    ## Define the api
    url <- parse_url(server)
    url$scheme <- "https"
    url$path<-file.path(workspace,"api", "v1", "interviews", intID, approve, fsep = "/")
    url$query <- list(comment = comment)
    ## authentication
    auth<-authenticate(apiUser, apiPass, type = "basic")

    server <- build_url(url)
    test_detail <- PATCH(url = server, auth)
    if(test_detail$status_code==406) {
        test_json<-data.table::data.table(resp_status=test_detail$status_code, intID=intID, description="incorrect status")
        return(test_json)
    } else if(test_detail$status_code==404) {
        test_json<-data.table::data.table(resp_status=test_detail$status_code, intID=intID, description="interview not found")
        return(test_json)

    } else if(test_detail$status_code==200) {
        test_json<-data.table::data.table(resp_status=test_detail$status_code, intID=intID, description="success")
        return(test_json)

    } else {
        test_json<-data.table::data.table(resp_status=test_detail$status_code, intID=intID, description="other error")
        return(test_json)
    }
}


